### DISCUSSION TITLE
Multiple Simultaneous connections & locking issues with extensions

### DISCUSSION THREAD

**Discussion by sevetseh28** [coder/code-server#6931] [Q&A]
Hi, I wanted to bring up a topic and ask if anyone has any tips or ideas on how to tackle this.

I have a Java environment with a Java project and all the Java extension pack installed.

The problem is that when I connect to the code-server, the Java Language extension starts building the project and performing all necessary Gradle steps to have it working. This process works well initially and as long as I am the only one connected to the Workspace, but as soon as someone else connects to the same server, the extension starts the build process again. This triggers my initial connection extension to restart the build again (unnecessarily), entering an infinite loop and creating deadlock conditions.

I’ve considered running multiple code-server processes with different users, while sharing the workspace, but I’m uncertain about this approach. I’m also unsure if code-server / VSCode Server was designed to support multiple simultaneous connections for pair-programming or if it’s just meant to be run by a single user.

Steps to Reproduce:
	1.	Connect to a code-server instance with a Java project and Java extension pack installed.
	2.	Observe the Java Language extension starting the build process.
	3.	Have a second user connect to the same code-server instance.
	4.	Observe the Java Language extension starting the build process again, causing the initial connection to restart the build process.

I appreciate any opinions or insights into the architecture of code-server that can help me tackle this problem.

Thank you!

**Comment by code-asher** [comment]
I think this would have to be solved by either:

1. Using separate copies of the project and separate instances of code-server.
2. Modifying the plugin itself to support running concurrently.
3. Modifying VS Code to share the extension host (likely a huge undertaking, but is probably the best).

It could also be interesting to see if Codespaces behaves any differently.  If it does the same thing, we can use that as leverage to report it upstream.

**Reply by code-asher** [reply]
I will also say that generally VS Code seems to be architectured to run for a single user, and pair programming is meant to be accomplished by using separate instances with a pair programming extension.

**Comment by sevetseh28** [comment]
Hey @code-asher  thanks so much for the reply. I'm gonna give it a try in Codespaces, I reckon the behaviour will be the same. I think Codespaces is relying on their proprietary Live Share ext for collaboration.
